File: ios/Sources/WattRouter/Transcript.swift
// Transcript.swift — turning a stream of events into something a person reads.
//
// History
//   2026-08-07  A. Sigdel  Created.
//
// Contents
//   TranscriptRow  One thing on screen.
//   Transcript     The rows so far, and the fold that produces them.
//
// A `TurnEvent` stream is not a transcript. Text arrives in fragments and belongs
// to whichever assistant turn is open; a call and its result are two events about
// one thing; a turn can fail after some of it has been shown. Folding that into
// rows is a state machine, and a state machine inside a `View` is one that can
// only be checked by looking at a screen.
//
// So this holds no SwiftUI and knows nothing about drawing. It is tested against
// event sequences, including the ones that are awkward rather than typical: text
// before any model has announced itself, a call whose result never arrives, two
// rounds of tools in one turn, and a failure partway that must leave what was
// already shown intact.

import Foundation

/// One thing on screen.
public enum TranscriptRow: Equatable, Sendable, Identifiable {
    /// What the person said.
    case said(id: Int, text: String)
    /// What a model answered. Open while text is still arriving.
    case answered(id: Int, model: String?, text: String)
    /// A tool the model asked for, and what it produced once it has.
    case used(id: Int, tool: String, result: String?)
    /// Why the turn stopped. Terminal for that turn, and never removed: a person
    /// who saw half an answer is owed the reason the rest never came.
    case failed(id: Int, reason: String)

    public var id: Int {
        switch self {
        case .said(let id, _), .answered(let id, _, _), .used(let id, _, _),
            .failed(let id, _):
            id
        }
    }
}

/// The rows so far.
///
/// # Atomic
/// Not thread-safe and not meant to be. One turn writes to it, from whichever
/// context is consuming the stream, and an interface reads it from the main
/// actor after each change.
public struct Transcript: Equatable, Sendable {
    /// Every row, oldest first.
    public private(set) var rows: [TranscriptRow] = []

    /// Row ids are positions in a sequence rather than in `rows`, so a row that
    /// changes keeps its identity and a list does not animate it as a new one.
    private var nextID = 0

    public init() {}

    /// Record something the person said, which also closes any open answer.
    ///
    /// A turn that is still streaming when the next thing is said has ended as
    /// far as the transcript is concerned; leaving it open would append the next
    /// model's text to the last one's paragraph.
    public mutating func said(_ text: String) {
        openAnswer = nil
        rows.append(.said(id: take(), text: text))
    }

    /// Fold one event in.
    public mutating func apply(_ event: TurnEvent) {
        switch event {
        case .answering(let model, _):
            // A model announcing itself opens an answer, or names one that text
            // already opened. `ChainWalk` sends this before any text, but a
            // fragment arriving first is not worth losing.
            if let index = openAnswer, case .answered(let id, _, let text) = rows[index] {
                rows[index] = .answered(id: id, model: model, text: text)
            } else {
                openAnswer = rows.count
                rows.append(.answered(id: take(), model: model, text: ""))
            }

        case .text(let fragment):
            if let index = openAnswer, case .answered(let id, let model, let text) = rows[index] {
                rows[index] = .answered(id: id, model: model, text: text + fragment)
            } else {
                openAnswer = rows.count
                rows.append(.answered(id: take(), model: nil, text: fragment))
            }

        case .toolCall(let call):
            // Closes the answer: text after a call belongs to the next round, and
            // appending it to the paragraph before would read as one thought.
            openAnswer = nil
            rows.append(.used(id: take(), tool: call.name, result: nil))

        case .toolResult(let result):
            // Fills in the earliest call still waiting. Tools run in order, so
            // the earliest unanswered call is the one this answers.
            if let index = rows.firstIndex(where: {
                if case .used(_, _, let filled) = $0 { return filled == nil }
                return false
            }), case .used(let id, let tool, _) = rows[index] {
                rows[index] = .used(id: id, tool: tool, result: result.content)
            }
        }
    }

    /// Record why the turn stopped, leaving everything already shown in place.
    public mutating func failed(_ reason: String) {
        openAnswer = nil
        rows.append(.failed(id: take(), reason: reason))
    }

    /// Index into `rows` of the answer still being written, if any.
    private var openAnswer: Int?

    private mutating func take() -> Int {
        defer { nextID += 1 }
        return nextID
    }
}

File: android/core/src/main/kotlin/com/getlora/wattrouter/Transcript.kt
// Transcript.kt — a turn's events, folded into what a person reads.
//
// History
//   2026-08-08  A. Sigdel  Created.
//
// Contents
//   Row         One line of the conversation as it is shown.
//   Transcript  The fold, and the rows it has produced.
//
// Every rule below exists because the naive version looks wrong on screen, and
// putting them in a composable would mean deciding per event whether to open a
// bubble or extend one — in a function that also has to lay one out.
//
// Ids come from a counter rather than a position in the list, so a row that
// grows keeps its identity. A LazyColumn keyed on position re-creates every
// item below an insert, which is visible as the whole conversation flickering
// each time a fragment arrives.

package com.getlora.wattrouter

/** One line of the conversation, as it is shown. */
sealed interface Row {
    val id: Int

    /** What the person typed. */
    data class Said(override val id: Int, val text: String) : Row

    /** What the model answered. [model] is null until one has announced itself. */
    data class Answered(override val id: Int, val model: String?, val text: String) : Row

    /** A tool, and its result once there is one. */
    data class Used(override val id: Int, val tool: String, val result: String?) : Row

    /** The turn stopped without answering. */
    data class Failed(override val id: Int, val reason: String) : Row

    /** The turn was interrupted. */
    data class Interrupted(override val id: Int) : Row
}

/** The conversation, as it is shown. */
class Transcript {
    private val backing = mutableListOf<Row>()

    /** In the order they happened. */
    val rows: List<Row> get() = backing

    private var next = 0

    /** Where the answer being written is, if one is open. */
    private var open: Int? = null

    /** A model that announced itself before saying anything. */
    private var pending: String? = null

    /** Record what the person typed. Closes any answer still open. */
    fun said(text: String) {
        open = null
        backing += Row.Said(next++, text)
    }

    /** Fold one event in. */
    fun apply(event: TurnEvent) {
        when (event) {
            // Names the row being written if there is one; otherwise waits. A
            // round that only calls tools would leave a blank bubble with a
            // model name on it.
            is TurnEvent.Answering -> open
                ?.let { at -> backing[at] = (backing[at] as Row.Answered).copy(model = event.model) }
                ?: run { pending = event.model }

            is TurnEvent.Text -> open
                ?.let { at ->
                    val row = backing[at] as Row.Answered
                    backing[at] = row.copy(text = row.text + event.text)
                }
                ?: run {
                    backing += Row.Answered(next++, pending, event.text)
                    pending = null
                    open = backing.lastIndex
                }

            // Closes the answer: anything the model says after a tool is a new
            // paragraph, after the result, not a continuation of the last one.
            is TurnEvent.Call -> {
                open = null
                backing += Row.Used(next++, event.call.name, result = null)
            }

            // The earliest unfilled one, which is valid because the Agent runs
            // tools in the order the model asked for them.
            is TurnEvent.Result -> {
                val at = backing.indexOfFirst { it is Row.Used && it.result == null }
                if (at >= 0) {
                    backing[at] = (backing[at] as Row.Used).copy(result = event.result.content)
                }
            }

            // No row. It feeds the routing panel, and "chose mid" in the middle
            // of a conversation is noise.
            is TurnEvent.Decided -> Unit
        }
    }

    /** Record that the turn was interrupted. */
    fun interrupted() {
        open = null
        pending = null
        backing += Row.Interrupted(next++)
    }

    /** Record that the turn stopped without answering. */
    fun failed(reason: String) {
        open = null
        pending = null
        backing += Row.Failed(next++, reason)
    }

    /**
     * Drop a trailing interruption, for a turn about to be resumed.
     *
     * @return whether there was one. A caller resumes only if so; otherwise it
     *   would be starting a turn nothing interrupted.
     */
    fun resumed(): Boolean {
        val last = backing.lastOrNull()
        if (last !is Row.Interrupted) return false
        backing.removeAt(backing.lastIndex)
        return true
    }
}
